x-pack/plugin/apm-integration/src/main/java/org/elasticsearch/xpack/apm/APM.java

```diff
@@ -44,7 +56,52 @@
         IndexNameExpressionResolver indexNameExpressionResolver,
         Supplier<RepositoriesService> repositoriesServiceSupplier
     ) {
-        tracer.set(new APMTracer());
+        tracer.set(new APMTracer(threadPool));
         return List.of(tracer.get());
     }
+
+    @Override
+    public List<TransportInterceptor> getTransportInterceptors(NamedWriteableRegistry namedWriteableRegistry, ThreadContext threadContext) {
+        return List.of(new TransportInterceptor() {
+            @Override
+            public AsyncSender interceptSender(AsyncSender sender) {
+                return new ApmTransportInterceptor(sender, threadContext);
+            }
+        });
+    }
+
+    private class ApmTransportInterceptor implements TransportInterceptor.AsyncSender {
+
+        private final TransportInterceptor.AsyncSender sender;
+        private final ThreadContext threadContext;
+
+        ApmTransportInterceptor(TransportInterceptor.AsyncSender sender, ThreadContext threadContext) {
+            this.sender = sender;
+            this.threadContext = threadContext;
+        }
+
+        @Override
+        public <T extends TransportResponse> void sendRequest(
+            Transport.Connection connection,
+            String action,
+            TransportRequest request,
+            TransportRequestOptions options,
+            TransportResponseHandler<T> handler
+        ) {
+            if (tracer.get() == null) {
+                sender.sendRequest(connection, action, request, options, handler);
+            } else {
+                var headers = tracer.get().getSpanHeadersByTaskId(request.getParentTask());
+                if (headers != null) {
+                    try (var context = threadContext.removeRequestHeaders(TRACE_HEADERS)) {
```

Comment:
Fixes an IDE warning about an unused variable:

```suggestion
                    try (var ignored = threadContext.removeRequestHeaders(TRACE_HEADERS)) {
```